### PR TITLE
Add first working example

### PR DESCRIPTION
**Pull request by MakisH**
This sets the structure of the repository and adds a first example. **Reviewers:** please comment on the structure and implementation of what is already there, we can always add more features later. Asking questions is highly recommended, as this is a new field for us.

The purpose of this repository is to provide code that can automatically build Virtual Machines with preCICE, adapters, solvers, and tutorials already installed. The target group is people that want to quickly try what preCICE can do, without having to install Linux or change anything else in their systems, as well as students that need a ready-to-use development environment for small projects.

See the README.md for details on what this does exactly.

### Why not directly distributing a VirtualBox image?

Directly distributing a pre-built image poses significant challenges regarding the size of the image and the target platforms. Distributing a specification makes it easier for us to maintain and easier for the user to upgrade, while it eliminates any license concerns.

Distributing a configuration also offers the advantage that the user understands how the system was built and they can reproduce or modify it. In any case, we can always export an image created with this system (e.g. for reproducibility reasons).

### Why not just a VirtualBox script?

We could use the VirtualBox command-line interface to accomplish the same task: create a VM, provision it with shell scripts, and potentially export it. However, the scripts would be specific to VirtualBox (there are more providers out there) and maybe also version-specific. Vagrant makes the specification more abstract, while it is an easy and cross-platform dependency.

### Why not Packer templates/images?

[Packer](https://www.packer.io/) is another tool from the same company as [Vagrant](https://www.vagrantup.com/), with their differences being a bit subtle to people not familiar enough with the respective field (myself included).

Vagrant downloads a "Vagrant base box", instructs a provider (such as VirtualBox) to create and start a virtual machine, sets up an SSH connection and a shared folder, and then executes provisioning steps (such as with shell scripts, Ansible, or other systems). It can also export the result into a new Vagrant base box (size in the scale of GB), to avoid any one-time provisioning steps. Its main purpose is to allow developers use a consistent, isolated environment for building a project.

Packer is meant to create such base boxes, not only for Vagrant, but also directly for Virtual Machine providers or even Docker. One can then provision these boxes with other tools (such as Vagrant, Ansible, or more) and then deploy them on some cloud provider with other tools (such as [Terraform](https://www.terraform.io/) from the same company). It starts e.g. from an `.iso` file of an operating system, sends keystrokes and other signals to install the OS, and so on. The result is an image (size in the scale of GB) and this is usually part of a CI pipeline.

I think that Packer would make sense if we had such a pipeline in place, which often updates images which we host somewhere. Both systems support versioning their produced artifacts. In any case, we would still need to decide if we want to produce Vagrant boxes or provider-specific boxes.

It feels to me that Packer is a more complex system that we currently don't need, while we can always easily refactor our system to incorporate it. Starting from an Ubuntu image is already close enough to what we need, as our provisioning steps are rather few (though maybe long due to compilation needs). Maybe worth looking at in the future, but I would need a clear motivation for that.

With either `vagrant package` or with Packer, we can [upload Vagrant boxes to the Vagrant Cloud](https://www.vagrantup.com/vagrant-cloud/boxes/create).

### Why provisioning with shell scripts and not with a configuration system?

Configuration systems such as [Ansible](https://www.ansible.com/) could simplify some configuration steps, as such a system would know the state of the VM. This is particularly important when provisioning (configuring) the system again after we start it: for example, we don't need to reinstall packages that we already installed, or we don't need to clone a repository a second time, so we first need to check if this steps has already been accomplished.

However, this would add additional dependencies, that are currently not well-motivated for our small-scale use case. It is definitely something worth looking at in the future, but no show-stopper.

### Why letting Vagrant run the provisioning scripts, instead of asking the user to execute them?

Asking the user to execute the script would simplify a few things on the technical side, offer some teaching benefits, and maybe allow for some flexibility of what to build. But the purpose here is that the user runs `vagrant up`, waits for the machine to be provisioned, and then is ready to run examples, without having to install anything else.

In any case, the user can simply comment-out some of the calls to the provisioning scripts to avoid building e.g. certain solvers.

## What is the goal after this is merged?

Next step would be to add more (if not all) officially-supported solvers/adapters. The structure implies that we can also add more platforms in the future, such as different Linux distributions (see Fedora), Windows, or macOS. The configuration-oriented scheme would make it possible for licensed users to get a VM with preCICE based on a proprietary OS.

### CODE REVIEW

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (fsimonis) [inline preCICE-Ubuntu-20.04/install-basics.sh:17]: We need to provide a simple and interactive way of changing this. Preferably an icon directly on the desktop.

File: preCICE-Ubuntu-20.04/install-openfoam.sh
Comment (fsimonis) [inline preCICE-Ubuntu-20.04/install-openfoam.sh:13]: This should not be necessary when using a login shell. See below.

File: preCICE-Ubuntu-20.04/install-openfoam.sh
Comment (fsimonis) [inline preCICE-Ubuntu-20.04/install-openfoam.sh:17]: Do we need the full repository?

Can we `wget -O - | tar zx ` an archive from github?
This is significantly faster.

Can we omit the history?
```suggestion
    sudo -u ${USER} git clone --depth=1 --branch master https://github.com/precice/openfoam-adapter.git
```

File: preCICE-Ubuntu-20.04/install-openfoam.sh
Comment (fsimonis) [inline preCICE-Ubuntu-20.04/install-openfoam.sh:16]: In case of reprovisioning the git repository, we could run `git pull` in the repo.

File: preCICE-Ubuntu-20.04/install-paraview.sh
Comment (fsimonis) [inline preCICE-Ubuntu-20.04/install-paraview.sh:10]: You could simplify this to [something like this](https://github.com/precice/precice/blob/22e061e7a2273ce7f2f84a1b930087d005aa6c29/tools/ci/travis-install-dependencies.sh#L64-L75).

File: preCICE-Ubuntu-20.04/install-precice.sh
Comment (fsimonis) [inline preCICE-Ubuntu-20.04/install-precice.sh:6]: Maybe using a debug build from source would be more helpful here because it allows to output debug log.
Nobody will notice the speed drop in the vm :wink:

File: preCICE-Ubuntu-20.04/install-openfoam.sh
Comment (fsimonis) [inline preCICE-Ubuntu-20.04/install-openfoam.sh:20]: Using an login shell may resolve the env issue.

```suggestion
sudo -u ${USER} -s bash -l ./Allwmake
```

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (fsimonis) [inline preCICE-Ubuntu-20.04/install-basics.sh:11]: Why not simply install the [minimal version of ubuntu desktop](https://packages.ubuntu.com/focal/ubuntu-desktop-minimal)?
This may be more familiar to users.

```
sudo apt-get install ubuntu-desktop-minimal
```

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-basics.sh:17]: The usual way would be that the user sets whatever they want (US, DE, Spanish, Italian, Greek, Klingon, ...) through the desktop environment tools. Most international keyboards follow the US layout, so I throught this is a good default. I think this belongs to an issue later.

File: preCICE-Ubuntu-20.04/install-openfoam.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-openfoam.sh]: Omiting the history is a good idea.

I still want the repository, as I would expect people to also do changes. I would like them to be able to contribute, even though then this becomes a question of target group.

File: preCICE-Ubuntu-20.04/install-openfoam.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-openfoam.sh:16]: Another good reason for keeping it a repository.

File: preCICE-Ubuntu-20.04/install-precice.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-precice.sh:6]: Right... I always forget that our packages do not support Debug mode...

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-basics.sh:11]: I will try, but I am afraid that it will have much higher memory requirements.

File: preCICE-Ubuntu-20.04/Vagrantfile
Comment (fsimonis) [inline preCICE-Ubuntu-20.04/Vagrantfile:16]: Not sure if the duplication works.

```suggestion
  # config.vm.synced_folder "../data", "/vagrant_data"
  # Disable the default folder sync
  config.vm.synced_folder ".", "/vagrant", disabled: true
  # Sync a dedicated shared folder
  config.vm.synced_folder "./shared", "/home/vagrant/Desktop/shared"
```

**Comment by fsimonis** [comment]
We can create a desktop shortcut to the "Region and Language" settings using this:
```
cp /usr/share/applications/gnome-region-panel.desktop /home/vagrant/Desktop/
chmod +x /home/vagrant/Desktop/gnome-region-panel.desktop
```

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-basics.sh:11]: This definitely looks more attractive, but I tried it and I have the following concerns:

* Memory usage: Xfce is around 300-400MB idle, Gnome is around 800-900MB idle (this is only a quick test I did).
* Gnome needs definitely more video memory (this may not be a real issue)
* I had very high latency in the GUI with a default `ubuntu-desktop-minimal`, I fixed them by changing the display controller to VMSVGA. I am not sure what the problem is exactly (I don't see it in other VMs), but it seems to me that this may cause some unexpected issues.
* With only `ubuntu-desktop-minimal` installed, the disk usage is 4.3GB. The current Xfce variant with preCICE, Python, OpenFOAM, ParaView, MPI, PETSc, ... is 10GB. Not directly comparable, but the GNOME image size may become a bit higher.

I would stick to Xfce for our first attempt. In any case, it is a fully functional desktop environment.

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (fsimonis) [reply preCICE-Ubuntu-20.04/install-basics.sh:17]: US as default is perfectly fine. More in an issue.

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (fsimonis) [reply preCICE-Ubuntu-20.04/install-basics.sh:11]: > I had very high latency in the GUI

Disabling animations makes the desktop more snappy. I didn't observe noteworthy lag though.

> I would stick to Xfce for our first attempt.

Sure! Let's stick to the resource friendly DE.

File: preCICE-Ubuntu-20.04/install-paraview.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-paraview.sh]: I updated this, but it is still not perfect. I would prefer to avoid version-specific strings anywhere else apart from the link.

File: preCICE-Ubuntu-20.04/install-precice.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-precice.sh]: Replaced this with preCICE built in Debug mode (build a Debian package and install it).

File: preCICE-Ubuntu-20.04/install-openfoam.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-openfoam.sh]: This worked like a charm, thanks!

File: preCICE-Ubuntu-20.04/Vagrantfile
Comment (MakisH) [reply preCICE-Ubuntu-20.04/Vagrantfile]: I don't see the reason to disable the `/vagrant/` part, but I like the explicit `~/Desktop/shared/`, I added it.

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (fsimonis) [reply preCICE-Ubuntu-20.04/install-basics.sh:13]: I tried LXDE and it is considerably more snappy in the VM. :+1:

File: preCICE-Ubuntu-20.04/install-openfoam.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-openfoam.sh:16]: I added a pull.

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-basics.sh:23]: I added a shortcut to the keyboard settings on the desktop.

**Comment by MakisH** [comment]
> We can create a desktop shortcut to the "Region and Language" settings using this:
> 
> ```
> cp /usr/share/applications/gnome-region-panel.desktop /home/vagrant/Desktop/
> chmod +x /home/vagrant/Desktop/gnome-region-panel.desktop
> ```

This is what I did, but for xfce it is:
```
/usr/share/applications/xfce-keyboard-settings.desktop
```

File: preCICE-Ubuntu-20.04/install-basics.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-basics.sh:13]: > Disabling animations makes the desktop more snappy. I didn't observe noteworthy lag though.

But this is again another tweak we don't need with xfce.

> I tried LXDE and it is considerably more snappy in the VM. 

LXDE is maybe the lightest, but also quite primitive (and meant as that). I find Xfce being on the sweet spot for the use case we have.

File: preCICE-Ubuntu-20.04/install-paraview.sh
Comment (MakisH) [reply preCICE-Ubuntu-20.04/install-paraview.sh]: Actually the `ParaView-5.8.1-MPI-Linux-Python3.7-64bit/` part is included in the ParaView archive, so I leave it like this for now.

**Comment by MakisH** [comment]
@fsimonis I applied all the review points. Do you maybe want to do a second review?

In a private discussion you suggested that we should not aim for multiple platforms. With this in mind, what if we move everything to the `precice/tutorials` repository? Then this would also make Vagrant's original use case happy:
```
git clone <tutorials>
vagrant up
```

**Comment by fsimonis** [comment]
> In a private discussion you suggested that we should not aim for multiple platforms.

That's right. A single platform is easier to maintain and support. Especially, when working in a live session such as a workshop. Providing multiple Linux distributions does not add any value and a Windows version will likely be too bloated.

> With this in mind, what if we move everything to the precice/tutorials repository? Then this would also make Vagrant's original use case happy:

This repository is supposed to provide some sort of preCICE sandbox for users. Using vagrant for the shared folder functionality sounds reasonable.
This user experience (let's call it the Extended Exploratory Experience of the Precise Code Interaction Coupling Environment) should live in a separate repository.
This allows us to keep the update cycles short and maybe even auto-generate and upload a ready-to-use vm.

I propose some final changes regarding the structure of the repository:
* Move the `shared` folder and the `Vagrantfile` to the root of the repository.
* Rename the folder `preCICE-Ubuntu-20.04` to `provisioning`.

Let's fix the structure and merge this ASAP and continue from there. This also allows us to document missing features in issues.

**Comment by MakisH** [comment]
@fsimonis I restructured the files, adapted the README, and renamed the VM to "preCICE-VM". I am ready to merge if you approve.

**Review by fsimonis — APPROVED** [review]
Let's get this merged.

**Pull request merged**
